File: app-laravel/src/Order/Application/Commands/UpdateOrderStatus/UpdateOrderStatusHandler.php
<?php

declare(strict_types=1);

namespace Src\Order\Application\Commands\UpdateOrderStatus;

use Src\Order\Domain\Repositories\OrderRepositoryInterface;
use Src\Order\Domain\ValueObjects\OrderId;
use Src\Order\Domain\ValueObjects\OrderStatus;
use Src\Shared\Application\Bus\Command;
use Src\Shared\Application\Bus\CommandHandler;
use Src\Shared\Domain\Exceptions\DomainException;

/**
 * UPDATE ORDER STATUS HANDLER
 * ============================
 * Sifariş statusunu yeniləyən handler.
 *
 * AXIN:
 * UpdateOrderStatusCommand
 *   → UpdateOrderStatusHandler
 *     → Repository-dən Order-i tap
 *     → Yeni statusu müəyyən et
 *     → Order-in müvafiq metodunu çağır (confirm, markAsPaid, ship, deliver, cancel)
 *     → Repository-yə saxla
 *
 * DİQQƏT: Hər status keçidi üçün Order entity-nin xüsusi metodu çağırılır.
 * Birbaşa status dəyişmək olmaz — bu State Machine qaydalarını qoruyur.
 */
class UpdateOrderStatusHandler implements CommandHandler
{
    public function __construct(
        private readonly OrderRepositoryInterface $orderRepository,
    ) {}

    /**
     * @param Command $command UpdateOrderStatusCommand olmalıdır
     * @throws DomainException Sifariş tapılmadıqda və ya keçid mümkün olmadıqda
     */
    public function handle(Command $command): void
    {
        /** @var UpdateOrderStatusCommand $command */

        // 1. Sifarişi DB-dən tap
        $order = $this->orderRepository->findById(
            OrderId::fromString($command->orderId()),
        );

        if ($order === null) {
            throw new DomainException("Sifariş tapılmadı: {$command->orderId()}");
        }

        // 2. Yeni statusu OrderStatus Value Object-ə çevir (validasiya olunur)
        $newStatus = new OrderStatus($command->newStatus());

        // 3. Hər status üçün müvafiq domain metodunu çağır
        // Birbaşa $order->setStatus() etmirik — bu encapsulation-ı pozar!
        // Hər metod öz biznes qaydalarını yoxlayır və event qeydə alır.
        match ($newStatus->value()) {
            OrderStatus::CONFIRMED => $order->confirm(),
            OrderStatus::PAID      => $order->markAsPaid(),
            OrderStatus::SHIPPED   => $order->ship(),
            OrderStatus::DELIVERED => $order->deliver(),
            OrderStatus::CANCELLED => $order->cancel('Admin tərəfindən ləğv edildi'),
            default => throw new DomainException(
                "'{$newStatus->value()}' statusuna keçid dəstəklənmir."
            ),
        };

        // 4. DB-yə saxla
        $this->orderRepository->save($order);
    }
}

File: app-spring/src/main/java/az/ecommerce/order/application/command/updateorderstatus/UpdateOrderStatusHandler.java
package az.ecommerce.order.application.command.updateorderstatus;

import az.ecommerce.order.domain.Order;
import az.ecommerce.order.domain.enums.OrderStatusEnum;
import az.ecommerce.order.domain.repository.OrderRepository;
import az.ecommerce.order.domain.valueobject.OrderId;
import az.ecommerce.shared.application.bus.CommandHandler;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import az.ecommerce.shared.infrastructure.bus.EventDispatcher;
import org.springframework.stereotype.Service;

@Service
public class UpdateOrderStatusHandler implements CommandHandler<UpdateOrderStatusCommand, Void> {

    private final OrderRepository repository;
    private final EventDispatcher eventDispatcher;

    public UpdateOrderStatusHandler(OrderRepository repository, EventDispatcher eventDispatcher) {
        this.repository = repository;
        this.eventDispatcher = eventDispatcher;
    }

    @Override
    public Void handle(UpdateOrderStatusCommand cmd) {
        Order order = repository.findById(new OrderId(cmd.orderId()))
                .orElseThrow(() -> new EntityNotFoundException("Order", cmd.orderId().toString()));

        switch (cmd.target()) {
            case CONFIRMED -> order.confirm();
            case PAID -> order.markAsPaid();
            case SHIPPED -> order.ship();
            case DELIVERED -> order.deliver();
            case CANCELLED -> order.cancel("Status keçidi ilə ləğv edildi");
            default -> throw new IllegalArgumentException("Bu status tetiklənə bilməz: " + cmd.target());
        }

        repository.save(order);
        eventDispatcher.dispatchAll(order);
        return null;
    }
}
